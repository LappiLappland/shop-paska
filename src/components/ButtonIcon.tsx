import { ReactNode, useRef } from 'react';
import useRipple from '../hooks/useRipple';

interface ButtonIconProps {
  children: ReactNode;
  text?: string;
  number?: number;
  size?: 'normal' | 'small';
  dataTestid?: string;
  ariaLabel?: string;
  onClick?: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
}

export default function ButtonIcon({
  children,
  text,
  number,
  size = 'normal',
  dataTestid,
  ariaLabel,
  onClick,
}: ButtonIconProps) {
  const containerEl = useRef<HTMLSpanElement>(null);
  const clickedEl = useRef<HTMLDivElement>(null);

  const shownNumber =
    !number || number < 1000
      ? number + ''
      : number.toString().slice(0, 3) + '+';

  const ripple = useRipple(containerEl, {
    clickRef: clickedEl,
    color: 'bg-primary/pressed',
    className: 'rounded-full',
  });

  const sizeClass =
    size === 'normal'
      ? ['w-10 h-10 md:w-12 md:h-12', 'w-6 h-6 md:w-8 md:h-8']
      : ['w-8 h-8', 'w-6 h-6'];

  return (
    <button
      className="group flex flex-col items-center justify-center"
      onClick={onClick}
      data-testid={dataTestid}
      aria-label={ariaLabel}
    >
      <div
        className={`relative ${sizeClass[0]} flex items-center justify-center`}
        ref={clickedEl}
      >
        <div
          className={`flex items-center justify-center ${sizeClass[1]} fill-on-surface-variant stroke-on-surface-variant group-hover:fill-primary group-hover:stroke-primary group-active:fill-primary/pressed group-active:stroke-primary/pressed`}
        >
          {children}
        </div>
        <span
          className={`
          absolute h-full w-full rounded-full
          group-hover:bg-primary/hovered-r
          group-active:bg-primary/pressed-r
          `}
          ref={containerEl}
        >
          {ripple}
        </span>
        <span
          className={`
        absolute -right-1.5 -top-1.5
        flex h-5 origin-center items-center justify-center rounded-full bg-error
        px-1.5 text-body-small
        font-medium
        text-on-error
        transition-transform
        duration-100
        ease-standart-acclerate
        ${number ? 'scale-100 text-on-error' : 'scale-0 text-on-error/0 delay-50'}
        `}
        >
          {shownNumber}
        </span>
      </div>
      <span className="text-label-large text-on-surface-variant group-hover:text-on-surface">
        {text}
      </span>
    </button>
  );
}
