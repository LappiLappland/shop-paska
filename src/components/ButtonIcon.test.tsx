import { act, render } from '@testing-library/react';
import ButtonIcon from './ButtonIcon';
import ArrowIcon from './icons/ArrowIcon';

describe('ButtonIcon component', () => {
  test('renders button icon with default size', () => {
    const { container } = render(
      <ButtonIcon>
        <ArrowIcon />
      </ButtonIcon>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders button icon with smaller size', () => {
    const { container } = render(
      <ButtonIcon size="small">
        <ArrowIcon />
      </ButtonIcon>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders button icon with different numbers', () => {
    const { container, rerender } = render(
      <ButtonIcon number={0}>
        <ArrowIcon />
      </ButtonIcon>
    );
    expect(container.firstChild).toMatchSnapshot();

    rerender(
      <ButtonIcon number={1}>
        <ArrowIcon />
      </ButtonIcon>
    );
    expect(container.firstChild).toMatchSnapshot();

    rerender(
      <ButtonIcon number={10}>
        <ArrowIcon />
      </ButtonIcon>
    );
    expect(container.firstChild).toMatchSnapshot();

    rerender(
      <ButtonIcon number={100}>
        <ArrowIcon />
      </ButtonIcon>
    );
    expect(container.firstChild).toMatchSnapshot();

    rerender(
      <ButtonIcon number={1000}>
        <ArrowIcon />
      </ButtonIcon>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('calls onClick handler when button icon is clicked', () => {
    const onClick = jest.fn();
    const { getByRole } = render(
      <ButtonIcon onClick={onClick}>
        <ArrowIcon />
      </ButtonIcon>
    );
    const buttonIcon = getByRole('button');
    act(() => {
      buttonIcon.click();
    })
    expect(onClick).toHaveBeenCalled();
  });
});
