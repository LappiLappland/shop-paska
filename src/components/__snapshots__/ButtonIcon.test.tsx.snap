// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonIcon component renders button icon with default size 1`] = `
<button
  class="group flex flex-col items-center justify-center"
>
  <div
    class="relative w-10 h-10 md:w-12 md:h-12 flex items-center justify-center"
  >
    <div
      class="flex items-center justify-center w-6 h-6 md:w-8 md:h-8 fill-on-surface-variant stroke-on-surface-variant group-hover:fill-primary group-hover:stroke-primary group-active:fill-primary/pressed group-active:stroke-primary/pressed"
    >
      <svg
        class="  "
        height="24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.293 4.707 14.586 12l-7.293 7.293 1.414 1.414L17.414 12 8.707 3.293 7.293 4.707z"
        />
      </svg>
    </div>
    <span
      class="
          absolute h-full w-full rounded-full
          group-hover:bg-primary/hovered-r
          group-active:bg-primary/pressed-r
          "
    >
      <span
        class="absolute inset-0 h-full w-full overflow-hidden rounded-full"
      />
    </span>
    <span
      class="
        absolute -right-1.5 -top-1.5
        flex h-5 origin-center items-center justify-center rounded-full bg-error
        px-1.5 text-body-small
        font-medium
        text-on-error
        transition-transform
        duration-100
        ease-standart-acclerate
        scale-0 text-on-error/0
        "
    >
      undefined
    </span>
  </div>
  <span
    class="text-label-large text-on-surface-variant group-hover:text-on-surface"
  />
</button>
`;

exports[`ButtonIcon component renders button icon with different numbers 1`] = `
<button
  class="group flex flex-col items-center justify-center"
>
  <div
    class="relative w-10 h-10 md:w-12 md:h-12 flex items-center justify-center"
  >
    <div
      class="flex items-center justify-center w-6 h-6 md:w-8 md:h-8 fill-on-surface-variant stroke-on-surface-variant group-hover:fill-primary group-hover:stroke-primary group-active:fill-primary/pressed group-active:stroke-primary/pressed"
    >
      <svg
        class="  "
        height="24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.293 4.707 14.586 12l-7.293 7.293 1.414 1.414L17.414 12 8.707 3.293 7.293 4.707z"
        />
      </svg>
    </div>
    <span
      class="
          absolute h-full w-full rounded-full
          group-hover:bg-primary/hovered-r
          group-active:bg-primary/pressed-r
          "
    >
      <span
        class="absolute inset-0 h-full w-full overflow-hidden rounded-full"
      />
    </span>
    <span
      class="
        absolute -right-1.5 -top-1.5
        flex h-5 origin-center items-center justify-center rounded-full bg-error
        px-1.5 text-body-small
        font-medium
        text-on-error
        transition-transform
        duration-100
        ease-standart-acclerate
        scale-0 text-on-error/0
        "
    >
      0
    </span>
  </div>
  <span
    class="text-label-large text-on-surface-variant group-hover:text-on-surface"
  />
</button>
`;

exports[`ButtonIcon component renders button icon with different numbers 2`] = `
<button
  class="group flex flex-col items-center justify-center"
>
  <div
    class="relative w-10 h-10 md:w-12 md:h-12 flex items-center justify-center"
  >
    <div
      class="flex items-center justify-center w-6 h-6 md:w-8 md:h-8 fill-on-surface-variant stroke-on-surface-variant group-hover:fill-primary group-hover:stroke-primary group-active:fill-primary/pressed group-active:stroke-primary/pressed"
    >
      <svg
        class="  "
        height="24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.293 4.707 14.586 12l-7.293 7.293 1.414 1.414L17.414 12 8.707 3.293 7.293 4.707z"
        />
      </svg>
    </div>
    <span
      class="
          absolute h-full w-full rounded-full
          group-hover:bg-primary/hovered-r
          group-active:bg-primary/pressed-r
          "
    >
      <span
        class="absolute inset-0 h-full w-full overflow-hidden rounded-full"
      />
    </span>
    <span
      class="
        absolute -right-1.5 -top-1.5
        flex h-5 origin-center items-center justify-center rounded-full bg-error
        px-1.5 text-body-small
        font-medium
        text-on-error
        transition-transform
        duration-100
        ease-standart-acclerate
        scale-100 text-on-error
        "
    >
      1
    </span>
  </div>
  <span
    class="text-label-large text-on-surface-variant group-hover:text-on-surface"
  />
</button>
`;

exports[`ButtonIcon component renders button icon with different numbers 3`] = `
<button
  class="group flex flex-col items-center justify-center"
>
  <div
    class="relative w-10 h-10 md:w-12 md:h-12 flex items-center justify-center"
  >
    <div
      class="flex items-center justify-center w-6 h-6 md:w-8 md:h-8 fill-on-surface-variant stroke-on-surface-variant group-hover:fill-primary group-hover:stroke-primary group-active:fill-primary/pressed group-active:stroke-primary/pressed"
    >
      <svg
        class="  "
        height="24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.293 4.707 14.586 12l-7.293 7.293 1.414 1.414L17.414 12 8.707 3.293 7.293 4.707z"
        />
      </svg>
    </div>
    <span
      class="
          absolute h-full w-full rounded-full
          group-hover:bg-primary/hovered-r
          group-active:bg-primary/pressed-r
          "
    >
      <span
        class="absolute inset-0 h-full w-full overflow-hidden rounded-full"
      />
    </span>
    <span
      class="
        absolute -right-1.5 -top-1.5
        flex h-5 origin-center items-center justify-center rounded-full bg-error
        px-1.5 text-body-small
        font-medium
        text-on-error
        transition-transform
        duration-100
        ease-standart-acclerate
        scale-100 text-on-error
        "
    >
      10
    </span>
  </div>
  <span
    class="text-label-large text-on-surface-variant group-hover:text-on-surface"
  />
</button>
`;

exports[`ButtonIcon component renders button icon with different numbers 4`] = `
<button
  class="group flex flex-col items-center justify-center"
>
  <div
    class="relative w-10 h-10 md:w-12 md:h-12 flex items-center justify-center"
  >
    <div
      class="flex items-center justify-center w-6 h-6 md:w-8 md:h-8 fill-on-surface-variant stroke-on-surface-variant group-hover:fill-primary group-hover:stroke-primary group-active:fill-primary/pressed group-active:stroke-primary/pressed"
    >
      <svg
        class="  "
        height="24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.293 4.707 14.586 12l-7.293 7.293 1.414 1.414L17.414 12 8.707 3.293 7.293 4.707z"
        />
      </svg>
    </div>
    <span
      class="
          absolute h-full w-full rounded-full
          group-hover:bg-primary/hovered-r
          group-active:bg-primary/pressed-r
          "
    >
      <span
        class="absolute inset-0 h-full w-full overflow-hidden rounded-full"
      />
    </span>
    <span
      class="
        absolute -right-1.5 -top-1.5
        flex h-5 origin-center items-center justify-center rounded-full bg-error
        px-1.5 text-body-small
        font-medium
        text-on-error
        transition-transform
        duration-100
        ease-standart-acclerate
        scale-100 text-on-error
        "
    >
      100
    </span>
  </div>
  <span
    class="text-label-large text-on-surface-variant group-hover:text-on-surface"
  />
</button>
`;

exports[`ButtonIcon component renders button icon with different numbers 5`] = `
<button
  class="group flex flex-col items-center justify-center"
>
  <div
    class="relative w-10 h-10 md:w-12 md:h-12 flex items-center justify-center"
  >
    <div
      class="flex items-center justify-center w-6 h-6 md:w-8 md:h-8 fill-on-surface-variant stroke-on-surface-variant group-hover:fill-primary group-hover:stroke-primary group-active:fill-primary/pressed group-active:stroke-primary/pressed"
    >
      <svg
        class="  "
        height="24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.293 4.707 14.586 12l-7.293 7.293 1.414 1.414L17.414 12 8.707 3.293 7.293 4.707z"
        />
      </svg>
    </div>
    <span
      class="
          absolute h-full w-full rounded-full
          group-hover:bg-primary/hovered-r
          group-active:bg-primary/pressed-r
          "
    >
      <span
        class="absolute inset-0 h-full w-full overflow-hidden rounded-full"
      />
    </span>
    <span
      class="
        absolute -right-1.5 -top-1.5
        flex h-5 origin-center items-center justify-center rounded-full bg-error
        px-1.5 text-body-small
        font-medium
        text-on-error
        transition-transform
        duration-100
        ease-standart-acclerate
        scale-100 text-on-error
        "
    >
      100+
    </span>
  </div>
  <span
    class="text-label-large text-on-surface-variant group-hover:text-on-surface"
  />
</button>
`;

exports[`ButtonIcon component renders button icon with smaller size 1`] = `
<button
  class="group flex flex-col items-center justify-center"
>
  <div
    class="relative w-8 h-8 flex items-center justify-center"
  >
    <div
      class="flex items-center justify-center w-6 h-6 fill-on-surface-variant stroke-on-surface-variant group-hover:fill-primary group-hover:stroke-primary group-active:fill-primary/pressed group-active:stroke-primary/pressed"
    >
      <svg
        class="  "
        height="24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.293 4.707 14.586 12l-7.293 7.293 1.414 1.414L17.414 12 8.707 3.293 7.293 4.707z"
        />
      </svg>
    </div>
    <span
      class="
          absolute h-full w-full rounded-full
          group-hover:bg-primary/hovered-r
          group-active:bg-primary/pressed-r
          "
    >
      <span
        class="absolute inset-0 h-full w-full overflow-hidden rounded-full"
      />
    </span>
    <span
      class="
        absolute -right-1.5 -top-1.5
        flex h-5 origin-center items-center justify-center rounded-full bg-error
        px-1.5 text-body-small
        font-medium
        text-on-error
        transition-transform
        duration-100
        ease-standart-acclerate
        scale-0 text-on-error/0
        "
    >
      undefined
    </span>
  </div>
  <span
    class="text-label-large text-on-surface-variant group-hover:text-on-surface"
  />
</button>
`;
